man/soapsuds.1
.\"
.\" soapsuds manual page.
.\" (C) 2003 Novell, Inc.
.\" Author:
.\"   Lluis Sanchez Gual (lluis@ximian.com)
.\"
.TH soapsuds 1
.SH NAME
soapsuds \- Mono's Remoting Proxy Generator
.SH SYNOPSIS
.PP
.B soapsuds
[options] [inputs] [outputs]
.SH DESCRIPTION
.I soapsuds
is a tool for generating WSDL documents and client proxies for remoting services.
A WSDL document describes the methods that a service provides together with XML schemas
of all referenced data structures. A client proxy is a class which offers the same methods
as the service it represents, but provides no implementation, it just forwards calls to the
remote service.
.PP
Generated client proxies are not strictly necessary, since the runtime can
automatically generate proxies from the original service types. However, it implies the need
of distributing the service class to all clients. In some scenarios a better approach is
to use soapsuds to generate a proxy class which has the same interface as the service, 
but without any implementation.
.PP
.I soapsuds 
can generate a WSDL document from a set of types specified with the option
.I -types
or can extract the types from an assembly specified with
.I -inputassemblyfile.
In this last case, soapsuds will include schemas for all serializable classes in 
the WSDL document.
.PP
Taking as input a WSDL document (or anything from which such document can be generated)
the tool can generate source code for the client proxy (using the
.I -gc
option) or it can directly generate an assembly that implements the proxy (with the
.I -outputassemblyfile
option). soapsuds will also generate fake serializable classes for all types defined
in the schema. However, beware that those classes will lack the semantics of the
original classes, since only data structure is generated.
.PP
.SH INPUT SOURCES
The following sources are available (only one can be specified):
.TP
.I "-urltoschema:url" "-url:url"
Url from which to retrieve the WSDL document.
.TP
.I "-types:type1,assembly[,serviceEndPoint][;type2,assembly[,serviceEndPoint]]..."
Specifies types for which to generate a schema or proxy. If no service endpoint is
provided, the one provided with the 
.I -serviceendpoint
option is used.
.TP
.I "-inputassemblyfile:file", "-ia:file"
Assembly that contains the types to export.
.TP
.I "-inputschemafile:schena", "-is:schema"
Schema from which to generate proxy classes.
.SH INPUT OPTIONS
The following options are available:
.TP
.I "-inputdirectory:directory", "-id:directory"
Directory where DLLs are located.
.TP
.I "-serviceendpoint:url", "-se:url"
Url of the service to be placed in the WSDL document.
.SH OUTPUTS
The following output options are available (more than one can be specified):
.TP
.I "-outputassemblyfile:file", "-oa:file"
Generates the proxy code and compiles it into an assembly with the provided name.
.TP
.I "-outputschemafile:file", "-os:file"
Generates a WSDL document that represents the provided types.
.TP
.I "-generatecode", "-gc"
Generates proxy source code for the provided WSDL document or types.
.SH OUTPUT OPTIONS
The following options are available:
.TP
.I "-outputdirectory:directory", "-od:directory"
Directory where generated files will be placed.
.TP
.I "-proxynamespace:namespace", "-pn:namespace"
Namespace of the generated proxy.
.TP
.I "-nowrappedproxy", "-nowp"
Do not generate a wrapped proxy.
.TP
.I "-wrappedproxy", "-wp"
Generate a wrapped proxy.
.TP
.I "-strongnamefile:file", "-sn:file"
Strong name file.
.SH GENERAL OPTIONS
The following options are available:
.TP
.I "-nologo"
Supress the startup logo.
.TP
.I "-u:username", "-user:username"
The user name to use when connecting to the server.
.TP
.I "-p:password", "-password:password"
The password to use when connecting to the server.
.TP
.I "-d:domain", "-domain:domain"
The domain to use when connecting to the server.
.TP
.I "-httpproxyname:name", "-hpn:name"
Name of the http proxy.
.TP
.I "-httpproxyport:port", "-hpp:port"
Port of the http proxy.
.PP
.SH AUTHORS
Lluis Sanchez Gual (lluis@ximian.com)
.PP
.SH LICENSE
wsdl is released under the terms of the GNU GPL.
.PP
.SH SEE ALSO
wsdl(1), disco(1), mono(1), mcs(1)


mcs/tools/soapsuds/soapsuds.cs
// 
// soapsuds.cs
//
// Author:
//   Lluis Sanchez Gual (lluis@ximian.com)
//
// Copyright (C) 2003 Ximian, Inc.
//

using System;
using System.Reflection;
using System.Collections;
using System.Runtime.Remoting.MetadataServices;
using System.Net;
using System.IO;

public class Driver
{
	static void Main (string[] args)
	{
		Runner run = new Runner (args);
		AppDomain domain = AppDomain.CreateDomain ("runner", null, Directory.GetCurrentDirectory(), "", false);
		domain.DoCallBack (new CrossAppDomainDelegate (run.Main));
	}
}

[Serializable]
class Runner
{
	static bool logo = true;
	static string inputUrl = null;
	static string inputTypes = null;
	static string inputSchema = null;
	static string inputAssembly = null;
	static string inputDirectory = null;

	static string serviceEndpoint = null;
	static string outputSchema = null;
	static string outputDirectory = null;
	static string outputAssembly = null;
	static bool outputCode = false;

	static bool wrappedProxy = true;
	static string proxyNamespace = null;
	static string strongNameFile = null;

	static string userName = null;
	static string password = null;
	static string domain = null;
	static string httpProxyName = null;
	static string httpProxyPort = null;
	
	string[] args;
	
	public Runner (string[] args)
	{
		this.args = args;
	}
	
	public void Main ()
	{
		try
		{
			ReadParameters (args);
			
			if (logo) 
				WriteLogo ();
			
			if (args.Length == 0 || args[0] == "--help")
			{
				WriteHelp ();
				return;
			}
			
			ArrayList types = new ArrayList ();
			Assembly assembly = null;
			
			if (inputAssembly != null)
			{
				assembly = Assembly.Load (inputAssembly);
				foreach (Type t in assembly.GetTypes ())
					types.Add (new ServiceType (t, serviceEndpoint));
			}
			
			if (inputTypes != null)
			{
				string[] ts = inputTypes.Split (';');
				foreach (string type in ts)
				{
					Type t = null;
					string url = null;

					string[] typeParts = type.Split (',');
					if (typeParts.Length == 1)
						throw new Exception ("Type assembly not specified");

					if (typeParts.Length >= 2)
					{
						t = Type.GetType (typeParts[0] + ", " + typeParts [1]);
						if (typeParts.Length > 2)
							url = typeParts [2];
					}
					types.Add (new ServiceType (t, url));
				}
			}
			
			ArrayList writtenFiles = new ArrayList ();
			MemoryStream schemaStream = null;
			
			if (types.Count > 0)
			{
				schemaStream = new MemoryStream ();
				MetaData.ConvertTypesToSchemaToStream ((ServiceType[]) types.ToArray (typeof(ServiceType)), SdlType.Wsdl, schemaStream);
			}
			
			if (inputUrl != null)
			{
				if (schemaStream != null) throw new Exception ("Only one type source can be specified");
				schemaStream = new MemoryStream ();
				MetaData.RetrieveSchemaFromUrlToStream (inputUrl, schemaStream);
			}
			
			if (inputSchema != null)
			{
				if (schemaStream != null) throw new Exception ("Only one type source can be specified");
				schemaStream = new MemoryStream ();
				
				FileStream fs = new FileStream (inputSchema, FileMode.Open, FileAccess.Read);
				byte[] buffer = new byte [1024*5];
				int nr = 0;
				while ((nr = fs.Read (buffer, 0, buffer.Length)) > 0)
					schemaStream.Write (buffer, 0, nr);
			}
			
			if (outputSchema != null)
			{
				if (schemaStream == null) throw new Exception ("No input schema or assembly has been specified");
				
				schemaStream.Position = 0;
				MetaData.SaveStreamToFile (schemaStream, outputSchema);
				Console.WriteLine ("Written file " + outputSchema);
			}
			
			if (outputCode)
			{
				if (schemaStream == null) throw new Exception ("No input schema or assembly has been specified");
				
				schemaStream.Position = 0;
				MetaData.ConvertSchemaStreamToCodeSourceStream (wrappedProxy, outputDirectory, schemaStream, writtenFiles, null, null);
			}
			
			if (outputAssembly != null)
			{
				if (schemaStream == null) throw new Exception ("No input schema or assembly has been specified");
				
				schemaStream.Position = 0;
				if (outputCode)
					MetaData.ConvertCodeSourceStreamToAssemblyFile (writtenFiles, outputAssembly, strongNameFile);
				else
				{
					MetaData.ConvertSchemaStreamToCodeSourceStream (wrappedProxy, outputDirectory, schemaStream, writtenFiles, null, null);
					MetaData.ConvertCodeSourceStreamToAssemblyFile (writtenFiles, outputAssembly, strongNameFile);
					foreach (string file in writtenFiles)
						File.Delete (file);
					writtenFiles.Clear ();
				}
				writtenFiles.Add (outputAssembly);
			}
			
			foreach (string fn in writtenFiles)
				Console.WriteLine ("Written file " + fn);
		}
		catch (Exception ex)
		{
			Console.WriteLine ("ERROR: " + ex.Message);
			if (ex.GetType() != typeof(Exception))
				Console.WriteLine (ex);
		}
		Console.WriteLine ();
	}

	static void WriteLogo ()
	{
		Console.WriteLine ("Mono SOAPSUDS Tool");
		Console.WriteLine ();
	}
	
	static void WriteHelp ()
	{
		Console.WriteLine ("Usage: soapsuds [inputs] [outputs] [options]");
		Console.WriteLine ();
		Console.WriteLine ("Inputs:");
		Console.WriteLine ("   -url urltoschema:url             Url from which to retrieve the schema");
		Console.WriteLine ("   -types:type1,assembly[,serviceEndpoint][;type2,assembly,...] ");
		Console.WriteLine ("                                    List of types from which to generate");
		Console.WriteLine ("                                    a schema or proxy");
		Console.WriteLine ("   -ia -inputassemblyfile:assembly  Assembly that contains the types to export");
		Console.WriteLine ("   -is -inputschemafile:schemafile  Schema from which to generate proxy classes");
		Console.WriteLine ();
		Console.WriteLine ("Input Options:");
		Console.WriteLine ("   -id -inputdirectory:directory    Directory where DLLs are located");
		Console.WriteLine ("   -se -serviceendpoint:url         Url of the service to be placed in the");
		Console.WriteLine ("                                    WSDL document");
		Console.WriteLine ();
		Console.WriteLine ("Outputs:");
		Console.WriteLine ("   -oa -outputassemblyfile:assembly Generate an assembly");
		Console.WriteLine ("   -os -outputschemafile:file       Generate a schema");
		Console.WriteLine ("   -gc -generatecode                Generate proxy source code");
		Console.WriteLine ();
		Console.WriteLine ("Output Options:");
		Console.WriteLine ("   -od -outputdirectory:directory   Directory where output will be generated");
		Console.WriteLine ("   -pn -proxynamespace:namespace    Namespace of the generated proxy");
		Console.WriteLine ("   -nowp -nowrappedproxy            Generate a wrapped proxy");
		Console.WriteLine ("   -wp -wrappedproxy                Generate a wrapped proxy");
		Console.WriteLine ("   -sn -strongnamefile:snfile       Strong name file");
		Console.WriteLine ();
		Console.WriteLine ("General Options:");
		Console.WriteLine ("   -u -username:name                User name for server authentication");
		Console.WriteLine ("   -p -password:pwd                 Password for server authentication");
		Console.WriteLine ("   -d -domain:domain                Domain of the server");
		Console.WriteLine ("   -hpn -httpProxyName:name         Name of http proxy");
		Console.WriteLine ("   -hpp -httpProxyPort:port         Port of http proxy");
		Console.WriteLine ("   -nologo                          Supress the startup logo");
		Console.WriteLine ();
	}
	
	static void ReadParameters (string[] args)
	{
		NetworkCredential cred = new NetworkCredential ();
		NetworkCredential proxyCred = new NetworkCredential ();
		WebProxy proxy = new WebProxy ();

		
		
		foreach (string arg in args)
		{
			if (!arg.StartsWith ("/") && !arg.StartsWith ("-"))
				continue;
			
			string parg = arg.Substring (1);
			int i = parg.IndexOf (":");
			string param = null;
			if (i != -1) {
				param = parg.Substring (i+1);
				parg = parg.Substring (0,i);
			}
			
			switch (parg.ToLower ())
			{
				case "nologo":
					logo = false;
					break;
					
				case "urltoschema": case "url":
					inputUrl = param;
					break;
					
				case "types":
					inputTypes = param;
					break;
					
				case "inputassemblyfile": case "ia":
					inputAssembly = param;
					break;
					
				case "outputassemblyfile": case "oa":
					outputAssembly = param;
					break;
					
				case "inputdirectory": case "id":
					inputDirectory = param;
					break;
					
				case "outputdirectory": case "od":
					outputDirectory = param;
					break;
					
				case "inputschemafile": case "is":
					inputSchema = param;
					break;
					
				case "outputschemafile": case "os":
					outputSchema = param;
					break;
					
				case "proxynamespace": case "pn":
					proxyNamespace = param;
					break;
					
				case "serviceendpoint": case "se":
					serviceEndpoint = param;
					break;
					
				case "strongnamefile": case "sn":
					strongNameFile = param;
					break;

				case "nowrappedproxy": case "nowp":
					wrappedProxy = false;
					break;
					
				case "wrappedproxy": case "wp":
					wrappedProxy = true;
					break;
					
				case "generatecode": case "gc":
					outputCode = true;
					break;
					
				case "username": case "u":
					userName = param;
					break;
					
				case "password": case "p":
					password = param;
					break;
					
				case "domain": case "d":
					domain = param;
					break;
					
				case "httpProxyName": case "hpn":
					httpProxyName = param;
					break;
					
				case "httpProxyPort": case "hpp":
					httpProxyPort = param;
					break;
			}
		}
	}
}
